src/core/default/mappers/simple/lambda_deployer.py

```diff
@@ -221,202 +222,381 @@ def _update_simple_lambda(
+        updated_configuration["Environment"] = {
+            "Variables": new_resource_environment_variables._d
+        }
 
-        mutable_previous_output["artifact_key"] = keyname
-        did_update_src_code = True
 
-    if (
-        not previous_resource.external_dependencies
-        == new_resource.external_dependencies
-    ):
-        if did_update_src_code:
-            output_task.update(
-                comment=f"[blink]Waiting for src code update to complete. ~5s[/blink]"
-            )
-            sleep(5)
+def _update_permissions(
+    output_task: OutputTask,
+    mutable_previous_output: Dict,
+    previous_output: Dict,
+    previous_resource: simple_xlambda.simple_function_model,
+    new_resource: simple_xlambda.simple_function_model,
+) -> bool:
+    if previous_resource.permissions == new_resource.permissions:
+        log.debug("Simple lambda, permissions didn't change")
+        return False
+
+    permission_output = previous_output.get("permissions")
+    role_name_output = previous_output.get("role_name")
+
+    permission_output = __update_permissions_create(
+        output_task,
+        role_name_output,
+        permission_output,
+        previous_resource,
+        new_resource,
+    )
+    permission_output = __update_permissions_remove(
+        output_task,
+        role_name_output,
+        permission_output,
+        previous_resource,
+        new_resource,
+    )
 
-        output_task.update(comment=f"Update Dependencies")
+    mutable_previous_output["permissions"] = permission_output
+    log.debug("Simple lambda, permissions updated")
+    return True
 
-        remove_dependencies = previous_resource.external_dependencies.difference(
-            new_resource.external_dependencies
-        )
-        create_dependencies = new_resource.external_dependencies.difference(
-            previous_resource.external_dependencies
+
+def __update_permissions_create(
+    output_task: OutputTask,
+    role_name_output: str,
+    permission_output: frozenset[dict],
+    previous_resource: simple_xlambda.simple_function_model,
+    new_resource: simple_xlambda.simple_function_model,
+) -> frozenset[Dict]:
+
+    create_permissions = new_resource.permissions.difference(
+        previous_resource.permissions
+    )
+    if not create_permissions:
+        log.debug("Simple lambda, no new permissions")
+        return permission_output
+
+    output_task.update(comment=f"Updating Policies, creating Permissions")
+    policies: List[frozendict] = []
+    for permission in create_permissions:
+        rv = frozendict(add_policy(role_name_output, permission))
+        policies.append(rv)
+
+    permission_output = frozenset(list(permission_output) + policies)
+    log.debug(f"Simple lambda, added {len(create_permissions)} permissions")
+    return permission_output
+
+
+def __update_permissions_remove(
+    output_task: OutputTask,
+    role_name_output: str,
+    permission_output: frozenset[dict],
```

Comment:
When running on my desktop with python 3.8.0, I am getting the following error 'Subscript for class "frozenset" will generate runtime exception; enclose type annotation in quotes (Pylance)'. Then whenever I run any command, it does throw a runtime exception.

This can be fixed by changing it to `'frozenset[dict]'`. I guess the quotes help turn it into a pure type annotation .

Reply:
Do we need a frozenset?

Reply:
It is a type hint on the args of the function, so it is good to have. Adding the quotes just signifies to Python to treat it as a pure type annotation as opposed to a Python object that supports type hints.